FirebaseMLModelDownloader/Sources/ModelInfoRetriever.swift

```diff
@@ -0,0 +1,251 @@
+    let bundleID = Bundle.main.bundleIdentifier!
+    request.setValue(bundleID, forHTTPHeaderField: ModelInfoRetriever.bundleIDHTTPHeader)
+    request.setValue(token, forHTTPHeaderField: ModelInfoRetriever.fisTokenHTTPHeader)
+    if let info = modelInfo, info.hash.count > 0 {
+      request.setValue(info.hash, forHTTPHeaderField: ModelInfoRetriever.hashMatchHTTPHeader)
+    }
+    return request
+  }
+
+  /// Get model info from server.
+  func downloadModelInfo(completion: @escaping (DownloadError?) -> Void) {
+    /// Get FIS token.
+    installations.authToken { [weak self] tokenResult, error in
+      guard let result = tokenResult
+      else { completion(.internalError(description: ModelInfoRetriever.tokenErrorDescription))
+        return
+      }
+      /// Get model info fetch URL with appropriate HTTP headers.
+      guard let request = self?.getModelInfoFetchURLRequest(token: result.authToken)
+      else {
+        completion(.internalError(description: ModelInfoRetriever.modelFetchURLErrorDescription))
+        return
+      }
+      /// Download model info.
+      let dataTask = URLSession.shared.dataTask(with: request) { [weak self]
+        data, response, error in
+        if let downloadError = error {
+          completion(.internalError(description: downloadError.localizedDescription))
+        } else {
+          guard let httpResponse = response as? HTTPURLResponse else {
+            completion(.internalError(description: ModelInfoRetriever
+                .invalidHTTPResponseErrorDescription))
+            return
+          }
+
+          guard httpResponse.statusCode == 200 || httpResponse.statusCode == 304 else {
+            // TODO: Improve http status code error handling
+            completion(.notFound)
+            return
+          }
+
+          /// Local model not modified.
+          if httpResponse.statusCode == 304 {
+            return
+          }
+
+          guard let modelHash = httpResponse
+            .allHeaderFields[ModelInfoRetriever.etagHTTPHeader] as? String else {
+            completion(.internalError(description: ModelInfoRetriever
+                .missingModelHashErrorDescription))
+            return
+          }
+
+          guard let data = data else {
+            completion(.internalError(description: ModelInfoRetriever
+                .invalidHTTPResponseErrorDescription))
+            return
+          }
+
+          self?.saveModelInfo(data: data, modelHash: modelHash)
+        }
+      }
+      dataTask.resume()
+    }
+  }
+
+  /// Save model info to user defaults.
+  func saveModelInfo(data: Data, modelHash: String) {
+    // TODO: Save model info to user defaults
+    modelInfo?.hash = modelHash
+  }
+}
+
+/// Named user defaults for FirebaseML.
+extension UserDefaults {
+  static var firebaseMLDefaults: UserDefaults {
+    let suiteName = "com.google.firebase.ml"
+    let defaults = UserDefaults(suiteName: suiteName)!
```

Comment:
Please avoid using force unwrap in the production code unless it is 100% safe or required as it potentially leads to unexpected obfuscated crashes.

Reply:
I looked into this and it seems like UserDefaults will only fail when you pass it a reserved (therefore invalid) suiteName. So technically this should work 100% of the time, but I've added a TODO here to look into this more and add a test in a future PR.

Reply:
From my experience, we should rely only on the public API definition as the implementation details may change without a notice, so I'm pretty convinced we should handle nullability gracefully here.

Reply:
Sounds good. I've added a TODO to handle this in a future PR.